### PR TITLE
feat: bundling the public key in the artifact for android

### PR DESCRIPTION
**Pull request by erickzanardo**
Adds logic to bundle a public key in the binary. Which afterwards can be used in the updater to validate patch signatures.

In this first POC we discussed to make the flow very simple. If there is a `public.pem` file in the root of the flutter project. It will be added, encoded in base64 into the "internal" `shorebird.yaml`.

iOS change will follow.

### CODE REVIEW

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 1]
```diff
@@ -1178,29 +1180,41 @@ class FlutterPlugin implements Plugin<Project> {
...
+                  def content = "";
+                  def outputDir = copyFlutterAssetsTask.destinationDir
+                  def publicKeyFile = new File("${project.rootDir.absolutePath}/../public.pem");
+
+                  def shorebirdYamlFile = new File("${outputDir}/flutter_assets/shorebird.yaml")
+                  if (publicKeyFile.exists()) {
+                    def base64EncodedKey = Base64.getEncoder().encode(publicKeyFile.readBytes());
+                    content += 'patch_public_key: ' + new String(base64EncodedKey) + '\n';
+                  }
+
```
Comment (eseidel) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: Is it possible for us to start to move our modifications here out into their own function?  I don't know much about gradle/groovy to know how practical that is (or if it's worth doing).

Comment (erickzanardo) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: yep! I guess that would work yes! I don't know much either, but I will figure it out

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 2]
```diff
@@ -1178,29 +1180,41 @@ class FlutterPlugin implements Plugin<Project> {
...
                 Task assembleTask = getAssembleTask(variant)
                 if (!shouldConfigureFlutterTask(assembleTask)) {
                   return
                 }
                 Task copyFlutterAssetsTask = addFlutterDeps(variant)
                 copyFlutterAssetsTask.doLast {
+
+                  def content = "";
+                  def outputDir = copyFlutterAssetsTask.destinationDir
+                  def publicKeyFile = new File("${project.rootDir.absolutePath}/../public.pem");
```
Comment (eseidel) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: This feels like it should be specified in shorebird.yaml, no?

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 3]
```diff
@@ -1178,29 +1180,41 @@ class FlutterPlugin implements Plugin<Project> {
...
+
+                  def shorebirdYamlFile = new File("${outputDir}/flutter_assets/shorebird.yaml")
+                  if (publicKeyFile.exists()) {
+                    def base64EncodedKey = Base64.getEncoder().encode(publicKeyFile.readBytes());
+                    content += 'patch_public_key: ' + new String(base64EncodedKey) + '\n';
+                  }
```
Comment (eseidel) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: I think the public key just ends up as a flutter asset?  Or is the idea to include the content of the key in the shorebird.yaml itself?

Comment (erickzanardo) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: The later, to include the key (as base64) into the "final" shorebird.yaml

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 4]
```diff
@@ -1178,29 +1180,41 @@ class FlutterPlugin implements Plugin<Project> {
         } // end def addFlutterDeps
 
         if (isFlutterAppProject()) {
+
```
Comment (eseidel) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]:
```suggestion
```

**Comment by eseidel** [comment]
Discussed in person, we're going to change this to expect an environment variable to have the key that we then just set in this shorebird.yaml.

e.g. essentially:

patch_public_key: $SHOREBIRD_PATCH_KEY

**Comment by eseidel** [comment]
We will then also have to change `shorebird` to do the base64 and pass it down through the ENV.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 5]
```diff
@@ -6,8 +6,8 @@ import com.android.build.OutputFile
 import groovy.json.JsonSlurper
 import groovy.json.JsonGenerator
 import groovy.xml.QName
+
 import java.nio.file.Paths
-import java.util.Set
```
Comment (erickzanardo) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:10]: Was unused

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 6]
```diff
@@ -28,7 +28,6 @@ import org.gradle.api.tasks.Optional
 import org.gradle.api.tasks.TaskAction
 import org.gradle.api.tasks.bundling.Jar
 import org.gradle.internal.os.OperatingSystem
-import org.gradle.util.VersionNumber
```
Comment (erickzanardo) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:31]: Was unused

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy (hunk 2 above)
Comment (erickzanardo) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy]: Like we discussed, changed to read from an env var instead

**Review by eseidel — APPROVED** [review]

**Pull request merged**